crates/forge/src/shared_cache.rs

```diff
@@ -0,0 +1,62 @@
+pub const CACHE_DIR: &str = ".snfoundry_cache";
+pub const PREV_TESTS_FAILED: &str = ".prev_tests_failed";
+
+use anyhow::{Ok, Result};
+use camino::Utf8PathBuf;
+use scarb_metadata::MetadataCommand;
+use std::fs::{self, File};
+use std::io::{BufRead, BufReader, BufWriter, Write};
+
+use crate::test_case_summary::TestCaseSummary;
+
+pub fn get_cached_failed_tests_names() -> Result<Option<Vec<String>>> {
+    let tests_failed_path = get_cache_dir()?.join(PREV_TESTS_FAILED);
+    if !tests_failed_path.exists() {
+        return Ok(None);
+    }
+
+    let file = File::open(tests_failed_path)?;
+    let buf: BufReader<File> = BufReader::new(file);
+    let tests: Vec<String> = buf
+        .lines()
+        .map(|l| l.expect("Could not parse line"))
+        .collect();
+    if tests.is_empty() {
+        return Ok(None);
+    }
+    Ok(Some(tests))
+}
+
+pub fn get_cache_dir() -> Result<Utf8PathBuf> {
+    let scarb_metadata = MetadataCommand::new().inherit_stderr().exec();
```

Comment:
I really dislike this hidden call to metadata. We should make sure to just use paths from the metadata read in `main_execution` (e.g. store path cache dir in `TestFilter`, prob not ideal solution though). The only case when this call is actually needed is for `clean_cache`